src/metabase/pulse/render/common.clj

```diff
@@ -25,55 +28,112 @@
-              (when (and currency? (= currency-style "code"))
-                (str (get-in currency/currency [(keyword (or currency "USD")) :code]) \space))
-              (.format fmtr (* value (or scale 1)))
-              (when (and currency? (= currency-style "name"))
-                (str \space (get-in currency/currency [(keyword (or currency "USD")) :name_plural])))
-              (when suffix suffix)))
+        (let [scaled-value (* value (or scale 1))
+              percent-scaled-value (* 100 scaled-value)
+              decimals-in-value (digits-after-decimal (if (= number-style "percent")
+                                                        (* 100 scaled-value)
+                                                        scaled-value))
+              decimal-digits (cond
+                               decimals decimals ;; if user ever specifies # of decimals, use that
+                               integral? 0
+                               currency? (get-in currency/currency [(keyword (or currency "USD")) :decimal_digits])
+                               (and percent? (> percent-scaled-value 100))   (min 2 decimals-in-value) ;; 5.5432 -> %554.32
+                               (and percent? (> 100 percent-scaled-value 1)) (min 0 decimals-in-value) ;; 0.2555 -> %26
+                               :else (if (>= scaled-value 1)
+                                       (min 2 decimals-in-value) ;; values greater than 1 round to 2 decimal places
+                                       (let [n-figs (sig-figs-after-decimal scaled-value decimal)]
+                                         (if (> n-figs 2)
+                                           (max 2 (- decimals-in-value (- n-figs 2))) ;; values less than 1 round to 2 sig-dig
+                                           decimals-in-value))))
+              fmt-str (cond-> base
+                        (not (zero? decimal-digits)) (str "." (apply str (repeat decimal-digits "0")))
+                        scientific? (str "E0")
+                        percent?    (str "%"))
+              fmtr (doto (DecimalFormat. fmt-str symbols) (.setRoundingMode RoundingMode/HALF_UP))]
+          (NumericWrapper.
+           (str (when prefix prefix)
+                (when (and currency? (or (nil? currency-style)
+                                         (= currency-style "symbol")))
+                  (get-in currency/currency [(keyword (or currency "USD")) :symbol]))
+                (when (and currency? (= currency-style "code"))
+                  (str (get-in currency/currency [(keyword (or currency "USD")) :code]) \space))
+                (cond-> (.format fmtr scaled-value)
+                  (not decimals) (strip-trailing-zeroes decimal))
+                (when (and currency? (= currency-style "name"))
+                  (str \space (get-in currency/currency [(keyword (or currency "USD")) :name_plural])))
+                (when suffix suffix))))
```

Comment:
I confess my eyes glazed over a bit, but I'll look at the unit tests more carefully